app/src/main/java/org/p2p/wallet/common/EncryptedSharedPreferences.kt

```diff
@@ -0,0 +1,255 @@
+package org.p2p.wallet.common
+
+import androidx.core.content.edit
+import android.content.SharedPreferences
+import com.google.gson.Gson
+import com.google.gson.reflect.TypeToken
+import timber.log.Timber
+import kotlin.properties.ReadWriteProperty
+import kotlin.reflect.KClass
+import kotlin.reflect.KProperty
+import org.p2p.core.crypto.Hex
+import org.p2p.wallet.common.crypto.keystore.DecodeCipher
+import org.p2p.wallet.common.crypto.keystore.EncodeCipher
+import org.p2p.wallet.common.crypto.keystore.KeyStoreWrapper
+
+class EncryptedSharedPreferences(
+    private val keyStoreWrapper: KeyStoreWrapper,
+    private val sharedPreferences: SharedPreferences,
+    private val gson: Gson
+) {
+    var loggingTag: String = "EncryptedSharedPreferences"
+
+    fun saveString(key: String, data: String) {
+        tryWithLog(key) {
+            val encodedData = keyStoreWrapper.encode(key, data)
+            sharedPreferences.edit { putString(key, encodedData) }
+        }
+    }
+
+    fun saveString(key: String, data: String, cipher: EncodeCipher) {
+        tryWithLog(key) {
+            val encodedData = keyStoreWrapper.encode(cipher, data)
+            sharedPreferences.edit { putString(key, encodedData) }
+        }
+    }
+
+    fun getString(key: String): String? {
+        return tryWithLog(key) {
+            val encodedData = sharedPreferences.getString(key, null)
+            encodedData?.let { keyStoreWrapper.decode(key, it) }
+        }
+    }
+
+    fun getString(key: String, cipher: DecodeCipher): String? {
+        return tryWithLog(key) {
+            val encodedData = sharedPreferences.getString(key, null)
+            encodedData?.let { keyStoreWrapper.decode(cipher, it) }
+        }
+    }
+
+    fun <T> saveObject(key: String, data: T) {
+        tryWithLog(key) {
+            val objectAsJson = gson.toJson(data)
+            val encodedData = keyStoreWrapper.encode(key, objectAsJson)
+            sharedPreferences.edit { putString(key, encodedData) }
+        }
+    }
+
+    /**
+     * !!! WARNING !!!
+     * DOESN'T WORK WITH CLASSES
+     */
+    @Deprecated(message = "DOESN'T WORK, use only with primitive types like String")
```

Comment:
If this class works only with primitives,
you can create something  like 
`@JvmInline
value class NonZeroInt(val value: Int) {
    init {
        require(value != 0) { "0 is not a value for this type!" }
    }
}
`

Reply:
I think annotate new method with @Deprecated annotation is not good idea